// @superfaceai/one-sdk/dist/internal/interpreter/variables.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.variablesToStrings = exports.getValue = exports.mergeVariables = exports.isEmptyRecord = exports.isNonPrimitive = exports.isPrimitive = exports.castToNonPrimitive = exports.castToVariables = exports.assertIsVariables = void 0;
var object_1 = require("../../lib/object");
var errors_1 = require("../errors");
function assertIsVariables(input) {
    if (!['string', 'number', 'boolean', 'object', 'undefined'].includes(typeof input)) {
        throw new errors_1.UnexpectedError("Invalid result type: " + typeof input);
    }
}
exports.assertIsVariables = assertIsVariables;
function castToVariables(input) {
    assertIsVariables(input);
    return input;
}
exports.castToVariables = castToVariables;
function castToNonPrimitive(input) {
    var variables = castToVariables(input);
    if (variables === undefined) {
        return undefined;
    }
    if (!isNonPrimitive(variables)) {
        throw new errors_1.UnexpectedError('Input is not NonPrimitive');
    }
    return variables;
}
exports.castToNonPrimitive = castToNonPrimitive;
function isPrimitive(input) {
    return (['string', 'number', 'boolean'].includes(typeof input) ||
        Array.isArray(input));
}
exports.isPrimitive = isPrimitive;
function isNonPrimitive(input) {
    return typeof input === 'object' && !Array.isArray(input);
}
exports.isNonPrimitive = isNonPrimitive;
function isEmptyRecord(input) {
    assertIsVariables(input);
    return isNonPrimitive(input) && Object.keys(input).length === 0;
}
exports.isEmptyRecord = isEmptyRecord;
/**
 * Recursively merges variables from `left` and then from `right` into a new object.
 */
var mergeVariables = function (left, right) {
    var e_1, _a, e_2, _b;
    var result = {};
    try {
        for (var _c = __values(Object.keys(left)), _d = _c.next(); !_d.done; _d = _c.next()) {
            var key = _d.value;
            result[key] = left[key];
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
        }
        finally { if (e_1) throw e_1.error; }
    }
    try {
        for (var _e = __values(Object.keys(right)), _f = _e.next(); !_f.done; _f = _e.next()) {
            var key = _f.value;
            var l = left[key];
            var r = right[key];
            if (r && l && isNonPrimitive(r) && isNonPrimitive(l)) {
                result[key] = exports.mergeVariables(l, r);
            }
            else {
                result[key] = right[key];
            }
        }
    }
    catch (e_2_1) { e_2 = { error: e_2_1 }; }
    finally {
        try {
            if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
        }
        finally { if (e_2) throw e_2.error; }
    }
    return result;
};
exports.mergeVariables = mergeVariables;
var getValue = function (variables, key) {
    if (variables === undefined) {
        return undefined;
    }
    var result;
    try {
        result = object_1.indexRecord(variables, key);
    }
    catch (_err) {
        // return undefined on error to preserve the original behavior of this function
        return undefined;
    }
    // sanity check, but if the input `variables` is correct then the result will be also
    assertIsVariables(result);
    return result;
};
exports.getValue = getValue;
/**
 * Stringifies a Record of variables. `undefined` values are removed.
 */
var variablesToStrings = function (variables) {
    var e_3, _a;
    var result = {};
    if (variables) {
        try {
            for (var _b = __values(Object.entries(variables)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var _d = __read(_c.value, 2), key = _d[0], value = _d[1];
                if (value !== undefined) {
                    result[key] = typeof value === 'string' ? value : JSON.stringify(value);
                }
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_3) throw e_3.error; }
        }
    }
    return result;
};
exports.variablesToStrings = variablesToStrings;
