// @superfaceai/parser/dist/interpreter/profile-output.utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isStringStructure = exports.isNumberStructure = exports.isBooleanStructure = exports.isObjectStructure = exports.isListStructure = exports.isEnumStructure = exports.isPrimitiveStructure = exports.isUnionStructure = exports.isNonNullStructure = exports.isScalarStructure = void 0;
function isScalarStructure(structure) {
    return structure.kind === 'ScalarStructure';
}
exports.isScalarStructure = isScalarStructure;
function isNonNullStructure(structure) {
    return structure.kind === 'NonNullStructure';
}
exports.isNonNullStructure = isNonNullStructure;
function isUnionStructure(structure) {
    return structure.kind === 'UnionStructure';
}
exports.isUnionStructure = isUnionStructure;
function isPrimitiveStructure(structure) {
    return structure.kind === 'PrimitiveStructure';
}
exports.isPrimitiveStructure = isPrimitiveStructure;
function isEnumStructure(structure) {
    return structure.kind === 'EnumStructure';
}
exports.isEnumStructure = isEnumStructure;
function isListStructure(structure) {
    return structure.kind === 'ListStructure';
}
exports.isListStructure = isListStructure;
function isObjectStructure(structure) {
    return structure.kind === 'ObjectStructure';
}
exports.isObjectStructure = isObjectStructure;
function isBooleanStructure(structure) {
    return isPrimitiveStructure(structure) && structure.type === 'boolean';
}
exports.isBooleanStructure = isBooleanStructure;
function isNumberStructure(structure) {
    return isPrimitiveStructure(structure) && structure.type === 'number';
}
exports.isNumberStructure = isNumberStructure;
function isStringStructure(structure) {
    return isPrimitiveStructure(structure) && structure.type === 'string';
}
exports.isStringStructure = isStringStructure;
