// @superfaceai/ast/dist/interfaces/superjson/superjson.utils.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isDigestSecurityValues = exports.isBearerTokenSecurityValues = exports.isBasicAuthSecurityValues = exports.isApiKeySecurityValues = exports.assertSuperJsonDocument = exports.isFileURIString = exports.isVersionString = void 0;
var typescript_is_1 = require("typescript-is");
var error_1 = require("../../error");
var superjson_1 = require("./superjson");
function isVersionString(input) {
    return superjson_1.SEMVER_REGEX.test(input);
}
exports.isVersionString = isVersionString;
function isFileURIString(input) {
    return superjson_1.FILE_URI_REGEX.test(input);
}
exports.isFileURIString = isFileURIString;
var assertSuperJsonDocument = function (input) {
    var e_1, _a, e_2, _b, e_3, _c;
    var parsedInput;
    try {
        parsedInput = typescript_is_1.assertEquals(input, function (object) { var path = ["input"]; function _undefined(object) { ; if (object !== undefined)
            return { message: "validation failed at " + path.join(".") + ": expected undefined", path: path.slice(), reason: { type: "undefined" } };
        else
            return null; } function _string(object) { ; if (typeof object !== "string")
            return { message: "validation failed at " + path.join(".") + ": expected a string", path: path.slice(), reason: { type: "string" } };
        else
            return null; } function sa__string_ea_11(object) { ; if (!Array.isArray(object))
            return { message: "validation failed at " + path.join(".") + ": expected an array", path: path.slice(), reason: { type: "array" } }; for (var i = 0; i < object.length; i++) {
            path.push("[" + i + "]");
            var error = _string(object[i]);
            path.pop();
            if (error)
                return error;
        } return null; } function _null(object) { ; if (object !== null)
            return { message: "validation failed at " + path.join(".") + ": expected null", path: path.slice(), reason: { type: "null" } };
        else
            return null; } function su__undefined_sa__string_ea_11_11_11_eu(object) {
            var e_4, _a;
            var conditions = [_undefined, sa__string_ea_11];
            try {
                for (var conditions_1 = __values(conditions), conditions_1_1 = conditions_1.next(); !conditions_1_1.done; conditions_1_1 = conditions_1.next()) {
                    var condition = conditions_1_1.value;
                    var error = condition(object);
                    if (!error)
                        return null;
                }
            }
            catch (e_4_1) { e_4 = { error: e_4_1 }; }
            finally {
                try {
                    if (conditions_1_1 && !conditions_1_1.done && (_a = conditions_1.return)) _a.call(conditions_1);
                }
                finally { if (e_4) throw e_4.error; }
            }
            return { message: "validation failed at " + path.join(".") + ": there are no valid alternatives", path: path.slice(), reason: { type: "union" } };
        } function _unknown() { return null; } function _393(object) {
            var e_5, _a;
            ;
            if (typeof object !== "object" || object === null || Array.isArray(object))
                return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } };
            try {
                for (var _b = __values(Object.keys(object)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var key = _c.value;
                    path.push(key);
                    var error = _unknown(object[key]);
                    path.pop();
                    if (error)
                        return error;
                }
            }
            catch (e_5_1) { e_5 = { error: e_5_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_5) throw e_5.error; }
            }
            return null;
        } function su__undefined__393_eu(object) {
            var e_6, _a;
            var conditions = [_undefined, _393];
            try {
                for (var conditions_2 = __values(conditions), conditions_2_1 = conditions_2.next(); !conditions_2_1.done; conditions_2_1 = conditions_2.next()) {
                    var condition = conditions_2_1.value;
                    var error = condition(object);
                    if (!error)
                        return null;
                }
            }
            catch (e_6_1) { e_6 = { error: e_6_1 }; }
            finally {
                try {
                    if (conditions_2_1 && !conditions_2_1.done && (_a = conditions_2.return)) _a.call(conditions_2);
                }
                finally { if (e_6) throw e_6.error; }
            }
            return { message: "validation failed at " + path.join(".") + ": there are no valid alternatives", path: path.slice(), reason: { type: "union" } };
        } function _false(object) { ; if (object !== false)
            return { message: "validation failed at " + path.join(".") + ": expected false", path: path.slice(), reason: { type: "boolean-literal", value: false } };
        else
            return null; } function _true(object) { ; if (object !== true)
            return { message: "validation failed at " + path.join(".") + ": expected true", path: path.slice(), reason: { type: "boolean-literal", value: true } };
        else
            return null; } function su__undefined__15__17_eu(object) {
            var e_7, _a;
            var conditions = [_undefined, _false, _true];
            try {
                for (var conditions_3 = __values(conditions), conditions_3_1 = conditions_3.next(); !conditions_3_1.done; conditions_3_1 = conditions_3.next()) {
                    var condition = conditions_3_1.value;
                    var error = condition(object);
                    if (!error)
                        return null;
                }
            }
            catch (e_7_1) { e_7 = { error: e_7_1 }; }
            finally {
                try {
                    if (conditions_3_1 && !conditions_3_1.done && (_a = conditions_3.return)) _a.call(conditions_3);
                }
                finally { if (e_7) throw e_7.error; }
            }
            return { message: "validation failed at " + path.join(".") + ": there are no valid alternatives", path: path.slice(), reason: { type: "union" } };
        } function _392(object) { ; if (typeof object !== "object" || object === null || Array.isArray(object))
            return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } }; {
            if ("input" in object) {
                path.push("input");
                var error = su__undefined__393_eu(object["input"]);
                path.pop();
                if (error)
                    return error;
            }
        } {
            if ("providerFailover" in object) {
                path.push("providerFailover");
                var error = su__undefined__15__17_eu(object["providerFailover"]);
                path.pop();
                if (error)
                    return error;
            }
        } return null; } function _388(object) {
            var e_8, _a;
            ;
            if (typeof object !== "object" || object === null || Array.isArray(object))
                return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } };
            try {
                for (var _b = __values(Object.keys(object)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var key = _c.value;
                    path.push(key);
                    var error = _392(object[key]);
                    path.pop();
                    if (error)
                        return error;
                }
            }
            catch (e_8_1) { e_8 = { error: e_8_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_8) throw e_8.error; }
            }
            return null;
        } function su__undefined__388_eu(object) {
            var e_9, _a;
            var conditions = [_undefined, _388];
            try {
                for (var conditions_4 = __values(conditions), conditions_4_1 = conditions_4.next(); !conditions_4_1.done; conditions_4_1 = conditions_4.next()) {
                    var condition = conditions_4_1.value;
                    var error = condition(object);
                    if (!error)
                        return null;
                }
            }
            catch (e_9_1) { e_9 = { error: e_9_1 }; }
            finally {
                try {
                    if (conditions_4_1 && !conditions_4_1.done && (_a = conditions_4.return)) _a.call(conditions_4);
                }
                finally { if (e_9) throw e_9.error; }
            }
            return { message: "validation failed at " + path.join(".") + ": there are no valid alternatives", path: path.slice(), reason: { type: "union" } };
        } function _407(object) {
            var e_10, _a;
            ;
            if (typeof object !== "object" || object === null || Array.isArray(object))
                return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } };
            try {
                for (var _b = __values(Object.keys(object)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var key = _c.value;
                    path.push(key);
                    var error = _unknown(object[key]);
                    path.pop();
                    if (error)
                        return error;
                }
            }
            catch (e_10_1) { e_10 = { error: e_10_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_10) throw e_10.error; }
            }
            return null;
        } function su__undefined__407_eu(object) {
            var e_11, _a;
            var conditions = [_undefined, _407];
            try {
                for (var conditions_5 = __values(conditions), conditions_5_1 = conditions_5.next(); !conditions_5_1.done; conditions_5_1 = conditions_5.next()) {
                    var condition = conditions_5_1.value;
                    var error = condition(object);
                    if (!error)
                        return null;
                }
            }
            catch (e_11_1) { e_11 = { error: e_11_1 }; }
            finally {
                try {
                    if (conditions_5_1 && !conditions_5_1.done && (_a = conditions_5.return)) _a.call(conditions_5);
                }
                finally { if (e_11) throw e_11.error; }
            }
            return { message: "validation failed at " + path.join(".") + ": there are no valid alternatives", path: path.slice(), reason: { type: "union" } };
        } function _409(object) { ; if (object !== "none")
            return { message: "validation failed at " + path.join(".") + ": expected string 'none'", path: path.slice(), reason: { type: "string-literal", value: "none" } };
        else
            return null; } function _411(object) { ; if (object !== "circuit-breaker")
            return { message: "validation failed at " + path.join(".") + ": expected string 'circuit-breaker'", path: path.slice(), reason: { type: "string-literal", value: "circuit-breaker" } };
        else
            return null; } function _414(object) { ; if (typeof object !== "object" || object === null || Array.isArray(object))
            return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } }; {
            if ("kind" in object) {
                path.push("kind");
                var error = _409(object["kind"]);
                path.pop();
                if (error)
                    return error;
            }
            else
                return { message: "validation failed at " + path.join(".") + ": expected 'kind' in object", path: path.slice(), reason: { type: "missing-property", property: "kind" } };
        } return null; } function _number(object) { ; if (typeof object !== "number")
            return { message: "validation failed at " + path.join(".") + ": expected a number", path: path.slice(), reason: { type: "number" } };
        else
            return null; } function su__undefined__number_eu(object) {
            var e_12, _a;
            var conditions = [_undefined, _number];
            try {
                for (var conditions_6 = __values(conditions), conditions_6_1 = conditions_6.next(); !conditions_6_1.done; conditions_6_1 = conditions_6.next()) {
                    var condition = conditions_6_1.value;
                    var error = condition(object);
                    if (!error)
                        return null;
                }
            }
            catch (e_12_1) { e_12 = { error: e_12_1 }; }
            finally {
                try {
                    if (conditions_6_1 && !conditions_6_1.done && (_a = conditions_6.return)) _a.call(conditions_6);
                }
                finally { if (e_12) throw e_12.error; }
            }
            return { message: "validation failed at " + path.join(".") + ": there are no valid alternatives", path: path.slice(), reason: { type: "union" } };
        } function _418(object) { ; if (object !== "exponential")
            return { message: "validation failed at " + path.join(".") + ": expected string 'exponential'", path: path.slice(), reason: { type: "string-literal", value: "exponential" } };
        else
            return null; } function _420(object) { ; if (typeof object !== "object" || object === null || Array.isArray(object))
            return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } }; {
            if ("kind" in object) {
                path.push("kind");
                var error = _418(object["kind"]);
                path.pop();
                if (error)
                    return error;
            }
            else
                return { message: "validation failed at " + path.join(".") + ": expected 'kind' in object", path: path.slice(), reason: { type: "missing-property", property: "kind" } };
        } {
            if ("start" in object) {
                path.push("start");
                var error = su__undefined__number_eu(object["start"]);
                path.pop();
                if (error)
                    return error;
            }
        } {
            if ("factor" in object) {
                path.push("factor");
                var error = su__undefined__number_eu(object["factor"]);
                path.pop();
                if (error)
                    return error;
            }
        } return null; } function su__undefined__418__420_eu(object) {
            var e_13, _a;
            var conditions = [_undefined, _418, _420];
            try {
                for (var conditions_7 = __values(conditions), conditions_7_1 = conditions_7.next(); !conditions_7_1.done; conditions_7_1 = conditions_7.next()) {
                    var condition = conditions_7_1.value;
                    var error = condition(object);
                    if (!error)
                        return null;
                }
            }
            catch (e_13_1) { e_13 = { error: e_13_1 }; }
            finally {
                try {
                    if (conditions_7_1 && !conditions_7_1.done && (_a = conditions_7.return)) _a.call(conditions_7);
                }
                finally { if (e_13) throw e_13.error; }
            }
            return { message: "validation failed at " + path.join(".") + ": there are no valid alternatives", path: path.slice(), reason: { type: "union" } };
        } function _415(object) { ; if (typeof object !== "object" || object === null || Array.isArray(object))
            return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } }; {
            if ("kind" in object) {
                path.push("kind");
                var error = _411(object["kind"]);
                path.pop();
                if (error)
                    return error;
            }
            else
                return { message: "validation failed at " + path.join(".") + ": expected 'kind' in object", path: path.slice(), reason: { type: "missing-property", property: "kind" } };
        } {
            if ("maxContiguousRetries" in object) {
                path.push("maxContiguousRetries");
                var error = su__undefined__number_eu(object["maxContiguousRetries"]);
                path.pop();
                if (error)
                    return error;
            }
        } {
            if ("requestTimeout" in object) {
                path.push("requestTimeout");
                var error = su__undefined__number_eu(object["requestTimeout"]);
                path.pop();
                if (error)
                    return error;
            }
        } {
            if ("backoff" in object) {
                path.push("backoff");
                var error = su__undefined__418__420_eu(object["backoff"]);
                path.pop();
                if (error)
                    return error;
            }
        } return null; } function su__undefined__409__411__414__415_eu(object) {
            var e_14, _a;
            var conditions = [_undefined, _409, _411, _414, _415];
            try {
                for (var conditions_8 = __values(conditions), conditions_8_1 = conditions_8.next(); !conditions_8_1.done; conditions_8_1 = conditions_8.next()) {
                    var condition = conditions_8_1.value;
                    var error = condition(object);
                    if (!error)
                        return null;
                }
            }
            catch (e_14_1) { e_14 = { error: e_14_1 }; }
            finally {
                try {
                    if (conditions_8_1 && !conditions_8_1.done && (_a = conditions_8.return)) _a.call(conditions_8);
                }
                finally { if (e_14) throw e_14.error; }
            }
            return { message: "validation failed at " + path.join(".") + ": there are no valid alternatives", path: path.slice(), reason: { type: "union" } };
        } function _406(object) { ; if (typeof object !== "object" || object === null || Array.isArray(object))
            return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } }; {
            if ("input" in object) {
                path.push("input");
                var error = su__undefined__407_eu(object["input"]);
                path.pop();
                if (error)
                    return error;
            }
        } {
            if ("retryPolicy" in object) {
                path.push("retryPolicy");
                var error = su__undefined__409__411__414__415_eu(object["retryPolicy"]);
                path.pop();
                if (error)
                    return error;
            }
        } return null; } function _404(object) {
            var e_15, _a;
            ;
            if (typeof object !== "object" || object === null || Array.isArray(object))
                return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } };
            try {
                for (var _b = __values(Object.keys(object)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var key = _c.value;
                    path.push(key);
                    var error = _406(object[key]);
                    path.pop();
                    if (error)
                        return error;
                }
            }
            catch (e_15_1) { e_15 = { error: e_15_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_15) throw e_15.error; }
            }
            return null;
        } function su__undefined__404_eu(object) {
            var e_16, _a;
            var conditions = [_undefined, _404];
            try {
                for (var conditions_9 = __values(conditions), conditions_9_1 = conditions_9.next(); !conditions_9_1.done; conditions_9_1 = conditions_9.next()) {
                    var condition = conditions_9_1.value;
                    var error = condition(object);
                    if (!error)
                        return null;
                }
            }
            catch (e_16_1) { e_16 = { error: e_16_1 }; }
            finally {
                try {
                    if (conditions_9_1 && !conditions_9_1.done && (_a = conditions_9.return)) _a.call(conditions_9);
                }
                finally { if (e_16) throw e_16.error; }
            }
            return { message: "validation failed at " + path.join(".") + ": there are no valid alternatives", path: path.slice(), reason: { type: "union" } };
        } function _396(object) { ; if (typeof object !== "object" || object === null || Array.isArray(object))
            return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } }; {
            if ("defaults" in object) {
                path.push("defaults");
                var error = su__undefined__404_eu(object["defaults"]);
                path.pop();
                if (error)
                    return error;
            }
        } return null; } function _397(object) { ; if (typeof object !== "object" || object === null || Array.isArray(object))
            return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } }; {
            if ("file" in object) {
                path.push("file");
                var error = _string(object["file"]);
                path.pop();
                if (error)
                    return error;
            }
            else
                return { message: "validation failed at " + path.join(".") + ": expected 'file' in object", path: path.slice(), reason: { type: "missing-property", property: "file" } };
        } return null; } function si__396__397_ei(object) {
            var e_17, _a;
            var conditions = [_396, _397];
            try {
                for (var conditions_10 = __values(conditions), conditions_10_1 = conditions_10.next(); !conditions_10_1.done; conditions_10_1 = conditions_10.next()) {
                    var condition = conditions_10_1.value;
                    var error = condition(object);
                    if (error)
                        return error;
                }
            }
            catch (e_17_1) { e_17 = { error: e_17_1 }; }
            finally {
                try {
                    if (conditions_10_1 && !conditions_10_1.done && (_a = conditions_10.return)) _a.call(conditions_10);
                }
                finally { if (e_17) throw e_17.error; }
            }
            return null;
        } function su__undefined__string_eu(object) {
            var e_18, _a;
            var conditions = [_undefined, _string];
            try {
                for (var conditions_11 = __values(conditions), conditions_11_1 = conditions_11.next(); !conditions_11_1.done; conditions_11_1 = conditions_11.next()) {
                    var condition = conditions_11_1.value;
                    var error = condition(object);
                    if (!error)
                        return null;
                }
            }
            catch (e_18_1) { e_18 = { error: e_18_1 }; }
            finally {
                try {
                    if (conditions_11_1 && !conditions_11_1.done && (_a = conditions_11.return)) _a.call(conditions_11);
                }
                finally { if (e_18) throw e_18.error; }
            }
            return { message: "validation failed at " + path.join(".") + ": there are no valid alternatives", path: path.slice(), reason: { type: "union" } };
        } function _398(object) { ; if (typeof object !== "object" || object === null || Array.isArray(object))
            return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } }; {
            if ("mapVariant" in object) {
                path.push("mapVariant");
                var error = su__undefined__string_eu(object["mapVariant"]);
                path.pop();
                if (error)
                    return error;
            }
        } {
            if ("mapRevision" in object) {
                path.push("mapRevision");
                var error = su__undefined__string_eu(object["mapRevision"]);
                path.pop();
                if (error)
                    return error;
            }
        } return null; } function si__396__398_ei(object) {
            var e_19, _a;
            var conditions = [_396, _398];
            try {
                for (var conditions_12 = __values(conditions), conditions_12_1 = conditions_12.next(); !conditions_12_1.done; conditions_12_1 = conditions_12.next()) {
                    var condition = conditions_12_1.value;
                    var error = condition(object);
                    if (error)
                        return error;
                }
            }
            catch (e_19_1) { e_19 = { error: e_19_1 }; }
            finally {
                try {
                    if (conditions_12_1 && !conditions_12_1.done && (_a = conditions_12.return)) _a.call(conditions_12);
                }
                finally { if (e_19) throw e_19.error; }
            }
            return null;
        } function su__string_si__396__397_ei_si__396__398_ei_eu(object) {
            var e_20, _a;
            var conditions = [_string, si__396__397_ei, si__396__398_ei];
            try {
                for (var conditions_13 = __values(conditions), conditions_13_1 = conditions_13.next(); !conditions_13_1.done; conditions_13_1 = conditions_13.next()) {
                    var condition = conditions_13_1.value;
                    var error = condition(object);
                    if (!error)
                        return null;
                }
            }
            catch (e_20_1) { e_20 = { error: e_20_1 }; }
            finally {
                try {
                    if (conditions_13_1 && !conditions_13_1.done && (_a = conditions_13.return)) _a.call(conditions_13);
                }
                finally { if (e_20) throw e_20.error; }
            }
            return { message: "validation failed at " + path.join(".") + ": there are no valid alternatives", path: path.slice(), reason: { type: "union" } };
        } function _390(object) {
            var e_21, _a;
            ;
            if (typeof object !== "object" || object === null || Array.isArray(object))
                return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } };
            try {
                for (var _b = __values(Object.keys(object)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var key = _c.value;
                    path.push(key);
                    var error = su__string_si__396__397_ei_si__396__398_ei_eu(object[key]);
                    path.pop();
                    if (error)
                        return error;
                }
            }
            catch (e_21_1) { e_21 = { error: e_21_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_21) throw e_21.error; }
            }
            return null;
        } function su__undefined__390_eu(object) {
            var e_22, _a;
            var conditions = [_undefined, _390];
            try {
                for (var conditions_14 = __values(conditions), conditions_14_1 = conditions_14.next(); !conditions_14_1.done; conditions_14_1 = conditions_14.next()) {
                    var condition = conditions_14_1.value;
                    var error = condition(object);
                    if (!error)
                        return null;
                }
            }
            catch (e_22_1) { e_22 = { error: e_22_1 }; }
            finally {
                try {
                    if (conditions_14_1 && !conditions_14_1.done && (_a = conditions_14.return)) _a.call(conditions_14);
                }
                finally { if (e_22) throw e_22.error; }
            }
            return { message: "validation failed at " + path.join(".") + ": there are no valid alternatives", path: path.slice(), reason: { type: "union" } };
        } function _380(object) { ; if (typeof object !== "object" || object === null || Array.isArray(object))
            return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } }; {
            if ("priority" in object) {
                path.push("priority");
                var error = su__undefined_sa__string_ea_11_11_11_eu(object["priority"]);
                path.pop();
                if (error)
                    return error;
            }
        } {
            if ("defaults" in object) {
                path.push("defaults");
                var error = su__undefined__388_eu(object["defaults"]);
                path.pop();
                if (error)
                    return error;
            }
        } {
            if ("providers" in object) {
                path.push("providers");
                var error = su__undefined__390_eu(object["providers"]);
                path.pop();
                if (error)
                    return error;
            }
        } return null; } function _381(object) { ; if (typeof object !== "object" || object === null || Array.isArray(object))
            return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } }; {
            if ("version" in object) {
                path.push("version");
                var error = _string(object["version"]);
                path.pop();
                if (error)
                    return error;
            }
            else
                return { message: "validation failed at " + path.join(".") + ": expected 'version' in object", path: path.slice(), reason: { type: "missing-property", property: "version" } };
        } return null; } function si__380_s__381_s_ei_s(object) {
            var e_23, _a, e_24, _b;
            var conditions = [_380, _381];
            try {
                for (var conditions_15 = __values(conditions), conditions_15_1 = conditions_15.next(); !conditions_15_1.done; conditions_15_1 = conditions_15.next()) {
                    var condition = conditions_15_1.value;
                    var error = condition(object);
                    if (error)
                        return error;
                }
            }
            catch (e_23_1) { e_23 = { error: e_23_1 }; }
            finally {
                try {
                    if (conditions_15_1 && !conditions_15_1.done && (_a = conditions_15.return)) _a.call(conditions_15);
                }
                finally { if (e_23) throw e_23.error; }
            }
            try {
                for (var _c = __values(Object.keys(object)), _d = _c.next(); !_d.done; _d = _c.next()) {
                    var key = _d.value;
                    if (key !== "priority" && key !== "defaults" && key !== "providers" && key !== "version")
                        return { message: "validation failed at " + path.join(".") + ": " + ("superfluous property '" + key + "' in object"), path: path.slice(), reason: { type: "superfluous-property" } };
                }
            }
            catch (e_24_1) { e_24 = { error: e_24_1 }; }
            finally {
                try {
                    if (_d && !_d.done && (_b = _c.return)) _b.call(_c);
                }
                finally { if (e_24) throw e_24.error; }
            }
            return null;
        } function _382(object) { ; if (typeof object !== "object" || object === null || Array.isArray(object))
            return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } }; {
            if ("file" in object) {
                path.push("file");
                var error = _string(object["file"]);
                path.pop();
                if (error)
                    return error;
            }
            else
                return { message: "validation failed at " + path.join(".") + ": expected 'file' in object", path: path.slice(), reason: { type: "missing-property", property: "file" } };
        } return null; } function si__380_s__382_s_ei_s(object) {
            var e_25, _a, e_26, _b;
            var conditions = [_380, _382];
            try {
                for (var conditions_16 = __values(conditions), conditions_16_1 = conditions_16.next(); !conditions_16_1.done; conditions_16_1 = conditions_16.next()) {
                    var condition = conditions_16_1.value;
                    var error = condition(object);
                    if (error)
                        return error;
                }
            }
            catch (e_25_1) { e_25 = { error: e_25_1 }; }
            finally {
                try {
                    if (conditions_16_1 && !conditions_16_1.done && (_a = conditions_16.return)) _a.call(conditions_16);
                }
                finally { if (e_25) throw e_25.error; }
            }
            try {
                for (var _c = __values(Object.keys(object)), _d = _c.next(); !_d.done; _d = _c.next()) {
                    var key = _d.value;
                    if (key !== "priority" && key !== "defaults" && key !== "providers" && key !== "file")
                        return { message: "validation failed at " + path.join(".") + ": " + ("superfluous property '" + key + "' in object"), path: path.slice(), reason: { type: "superfluous-property" } };
                }
            }
            catch (e_26_1) { e_26 = { error: e_26_1 }; }
            finally {
                try {
                    if (_d && !_d.done && (_b = _c.return)) _b.call(_c);
                }
                finally { if (e_26) throw e_26.error; }
            }
            return null;
        } function su__string_si__380__381_ei_si__380__382_ei_eu(object) {
            var e_27, _a;
            var conditions = [_string, si__380_s__381_s_ei_s, si__380_s__382_s_ei_s];
            try {
                for (var conditions_17 = __values(conditions), conditions_17_1 = conditions_17.next(); !conditions_17_1.done; conditions_17_1 = conditions_17.next()) {
                    var condition = conditions_17_1.value;
                    var error = condition(object);
                    if (!error)
                        return null;
                }
            }
            catch (e_27_1) { e_27 = { error: e_27_1 }; }
            finally {
                try {
                    if (conditions_17_1 && !conditions_17_1.done && (_a = conditions_17.return)) _a.call(conditions_17);
                }
                finally { if (e_27) throw e_27.error; }
            }
            return { message: "validation failed at " + path.join(".") + ": there are no valid alternatives", path: path.slice(), reason: { type: "union" } };
        } function _376_s(object) {
            var e_28, _a;
            ;
            if (typeof object !== "object" || object === null || Array.isArray(object))
                return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } };
            try {
                for (var _b = __values(Object.keys(object)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var key = _c.value;
                    path.push(key);
                    var error = su__string_si__380__381_ei_si__380__382_ei_eu(object[key]);
                    path.pop();
                    if (error)
                        return error;
                }
            }
            catch (e_28_1) { e_28 = { error: e_28_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_28) throw e_28.error; }
            }
            return null;
        } function su__undefined__376_eu(object) {
            var e_29, _a;
            var conditions = [_undefined, _376_s];
            try {
                for (var conditions_18 = __values(conditions), conditions_18_1 = conditions_18.next(); !conditions_18_1.done; conditions_18_1 = conditions_18.next()) {
                    var condition = conditions_18_1.value;
                    var error = condition(object);
                    if (!error)
                        return null;
                }
            }
            catch (e_29_1) { e_29 = { error: e_29_1 }; }
            finally {
                try {
                    if (conditions_18_1 && !conditions_18_1.done && (_a = conditions_18.return)) _a.call(conditions_18);
                }
                finally { if (e_29) throw e_29.error; }
            }
            return { message: "validation failed at " + path.join(".") + ": there are no valid alternatives", path: path.slice(), reason: { type: "union" } };
        } function _424(object) { ; if (typeof object !== "object" || object === null || Array.isArray(object))
            return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } }; {
            if ("id" in object) {
                path.push("id");
                var error = _string(object["id"]);
                path.pop();
                if (error)
                    return error;
            }
            else
                return { message: "validation failed at " + path.join(".") + ": expected 'id' in object", path: path.slice(), reason: { type: "missing-property", property: "id" } };
        } return null; } function _425(object) { ; if (typeof object !== "object" || object === null || Array.isArray(object))
            return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } }; {
            if ("apikey" in object) {
                path.push("apikey");
                var error = _string(object["apikey"]);
                path.pop();
                if (error)
                    return error;
            }
            else
                return { message: "validation failed at " + path.join(".") + ": expected 'apikey' in object", path: path.slice(), reason: { type: "missing-property", property: "apikey" } };
        } return null; } function si__424_s__425_s_ei_s(object) {
            var e_30, _a, e_31, _b;
            var conditions = [_424, _425];
            try {
                for (var conditions_19 = __values(conditions), conditions_19_1 = conditions_19.next(); !conditions_19_1.done; conditions_19_1 = conditions_19.next()) {
                    var condition = conditions_19_1.value;
                    var error = condition(object);
                    if (error)
                        return error;
                }
            }
            catch (e_30_1) { e_30 = { error: e_30_1 }; }
            finally {
                try {
                    if (conditions_19_1 && !conditions_19_1.done && (_a = conditions_19.return)) _a.call(conditions_19);
                }
                finally { if (e_30) throw e_30.error; }
            }
            try {
                for (var _c = __values(Object.keys(object)), _d = _c.next(); !_d.done; _d = _c.next()) {
                    var key = _d.value;
                    if (key !== "id" && key !== "apikey")
                        return { message: "validation failed at " + path.join(".") + ": " + ("superfluous property '" + key + "' in object"), path: path.slice(), reason: { type: "superfluous-property" } };
                }
            }
            catch (e_31_1) { e_31 = { error: e_31_1 }; }
            finally {
                try {
                    if (_d && !_d.done && (_b = _c.return)) _b.call(_c);
                }
                finally { if (e_31) throw e_31.error; }
            }
            return null;
        } function _427(object) { ; if (typeof object !== "object" || object === null || Array.isArray(object))
            return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } }; {
            if ("username" in object) {
                path.push("username");
                var error = _string(object["username"]);
                path.pop();
                if (error)
                    return error;
            }
            else
                return { message: "validation failed at " + path.join(".") + ": expected 'username' in object", path: path.slice(), reason: { type: "missing-property", property: "username" } };
        } {
            if ("password" in object) {
                path.push("password");
                var error = _string(object["password"]);
                path.pop();
                if (error)
                    return error;
            }
            else
                return { message: "validation failed at " + path.join(".") + ": expected 'password' in object", path: path.slice(), reason: { type: "missing-property", property: "password" } };
        } return null; } function si__424_s__427_s_ei_s(object) {
            var e_32, _a, e_33, _b;
            var conditions = [_424, _427];
            try {
                for (var conditions_20 = __values(conditions), conditions_20_1 = conditions_20.next(); !conditions_20_1.done; conditions_20_1 = conditions_20.next()) {
                    var condition = conditions_20_1.value;
                    var error = condition(object);
                    if (error)
                        return error;
                }
            }
            catch (e_32_1) { e_32 = { error: e_32_1 }; }
            finally {
                try {
                    if (conditions_20_1 && !conditions_20_1.done && (_a = conditions_20.return)) _a.call(conditions_20);
                }
                finally { if (e_32) throw e_32.error; }
            }
            try {
                for (var _c = __values(Object.keys(object)), _d = _c.next(); !_d.done; _d = _c.next()) {
                    var key = _d.value;
                    if (key !== "id" && key !== "username" && key !== "password")
                        return { message: "validation failed at " + path.join(".") + ": " + ("superfluous property '" + key + "' in object"), path: path.slice(), reason: { type: "superfluous-property" } };
                }
            }
            catch (e_33_1) { e_33 = { error: e_33_1 }; }
            finally {
                try {
                    if (_d && !_d.done && (_b = _c.return)) _b.call(_c);
                }
                finally { if (e_33) throw e_33.error; }
            }
            return null;
        } function _429(object) { ; if (typeof object !== "object" || object === null || Array.isArray(object))
            return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } }; {
            if ("token" in object) {
                path.push("token");
                var error = _string(object["token"]);
                path.pop();
                if (error)
                    return error;
            }
            else
                return { message: "validation failed at " + path.join(".") + ": expected 'token' in object", path: path.slice(), reason: { type: "missing-property", property: "token" } };
        } return null; } function si__424_s__429_s_ei_s(object) {
            var e_34, _a, e_35, _b;
            var conditions = [_424, _429];
            try {
                for (var conditions_21 = __values(conditions), conditions_21_1 = conditions_21.next(); !conditions_21_1.done; conditions_21_1 = conditions_21.next()) {
                    var condition = conditions_21_1.value;
                    var error = condition(object);
                    if (error)
                        return error;
                }
            }
            catch (e_34_1) { e_34 = { error: e_34_1 }; }
            finally {
                try {
                    if (conditions_21_1 && !conditions_21_1.done && (_a = conditions_21.return)) _a.call(conditions_21);
                }
                finally { if (e_34) throw e_34.error; }
            }
            try {
                for (var _c = __values(Object.keys(object)), _d = _c.next(); !_d.done; _d = _c.next()) {
                    var key = _d.value;
                    if (key !== "id" && key !== "token")
                        return { message: "validation failed at " + path.join(".") + ": " + ("superfluous property '" + key + "' in object"), path: path.slice(), reason: { type: "superfluous-property" } };
                }
            }
            catch (e_35_1) { e_35 = { error: e_35_1 }; }
            finally {
                try {
                    if (_d && !_d.done && (_b = _c.return)) _b.call(_c);
                }
                finally { if (e_35) throw e_35.error; }
            }
            return null;
        } function _431(object) { ; if (typeof object !== "object" || object === null || Array.isArray(object))
            return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } }; {
            if ("digest" in object) {
                path.push("digest");
                var error = _string(object["digest"]);
                path.pop();
                if (error)
                    return error;
            }
            else
                return { message: "validation failed at " + path.join(".") + ": expected 'digest' in object", path: path.slice(), reason: { type: "missing-property", property: "digest" } };
        } return null; } function si__424_s__431_s_ei_s(object) {
            var e_36, _a, e_37, _b;
            var conditions = [_424, _431];
            try {
                for (var conditions_22 = __values(conditions), conditions_22_1 = conditions_22.next(); !conditions_22_1.done; conditions_22_1 = conditions_22.next()) {
                    var condition = conditions_22_1.value;
                    var error = condition(object);
                    if (error)
                        return error;
                }
            }
            catch (e_36_1) { e_36 = { error: e_36_1 }; }
            finally {
                try {
                    if (conditions_22_1 && !conditions_22_1.done && (_a = conditions_22.return)) _a.call(conditions_22);
                }
                finally { if (e_36) throw e_36.error; }
            }
            try {
                for (var _c = __values(Object.keys(object)), _d = _c.next(); !_d.done; _d = _c.next()) {
                    var key = _d.value;
                    if (key !== "id" && key !== "digest")
                        return { message: "validation failed at " + path.join(".") + ": " + ("superfluous property '" + key + "' in object"), path: path.slice(), reason: { type: "superfluous-property" } };
                }
            }
            catch (e_37_1) { e_37 = { error: e_37_1 }; }
            finally {
                try {
                    if (_d && !_d.done && (_b = _c.return)) _b.call(_c);
                }
                finally { if (e_37) throw e_37.error; }
            }
            return null;
        } function su_si__424__425_ei_si__424__427_ei_si__424__429_ei_si__424__431_ei_eu(object) {
            var e_38, _a;
            var conditions = [si__424_s__425_s_ei_s, si__424_s__427_s_ei_s, si__424_s__429_s_ei_s, si__424_s__431_s_ei_s];
            try {
                for (var conditions_23 = __values(conditions), conditions_23_1 = conditions_23.next(); !conditions_23_1.done; conditions_23_1 = conditions_23.next()) {
                    var condition = conditions_23_1.value;
                    var error = condition(object);
                    if (!error)
                        return null;
                }
            }
            catch (e_38_1) { e_38 = { error: e_38_1 }; }
            finally {
                try {
                    if (conditions_23_1 && !conditions_23_1.done && (_a = conditions_23.return)) _a.call(conditions_23);
                }
                finally { if (e_38) throw e_38.error; }
            }
            return { message: "validation failed at " + path.join(".") + ": there are no valid alternatives", path: path.slice(), reason: { type: "union" } };
        } function sa_su_si__424__425_ei_si__424__427_ei_si__424__429_ei_si__424__431_ei_eu_ea_433(object) { ; if (!Array.isArray(object))
            return { message: "validation failed at " + path.join(".") + ": expected an array", path: path.slice(), reason: { type: "array" } }; for (var i = 0; i < object.length; i++) {
            path.push("[" + i + "]");
            var error = su_si__424__425_ei_si__424__427_ei_si__424__429_ei_si__424__431_ei_eu(object[i]);
            path.pop();
            if (error)
                return error;
        } return null; } function su__undefined_sa_su_si__424__425_ei_si__424__427_ei_si__424__429_ei_si__424__431_ei_eu_ea_433_433_433_eu(object) {
            var e_39, _a;
            var conditions = [_undefined, sa_su_si__424__425_ei_si__424__427_ei_si__424__429_ei_si__424__431_ei_eu_ea_433];
            try {
                for (var conditions_24 = __values(conditions), conditions_24_1 = conditions_24.next(); !conditions_24_1.done; conditions_24_1 = conditions_24.next()) {
                    var condition = conditions_24_1.value;
                    var error = condition(object);
                    if (!error)
                        return null;
                }
            }
            catch (e_39_1) { e_39 = { error: e_39_1 }; }
            finally {
                try {
                    if (conditions_24_1 && !conditions_24_1.done && (_a = conditions_24.return)) _a.call(conditions_24);
                }
                finally { if (e_39) throw e_39.error; }
            }
            return { message: "validation failed at " + path.join(".") + ": there are no valid alternatives", path: path.slice(), reason: { type: "union" } };
        } function _436_s(object) {
            var e_40, _a;
            ;
            if (typeof object !== "object" || object === null || Array.isArray(object))
                return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } };
            try {
                for (var _b = __values(Object.keys(object)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var key = _c.value;
                    path.push(key);
                    var error = _string(object[key]);
                    path.pop();
                    if (error)
                        return error;
                }
            }
            catch (e_40_1) { e_40 = { error: e_40_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_40) throw e_40.error; }
            }
            return null;
        } function su__undefined__436_eu(object) {
            var e_41, _a;
            var conditions = [_undefined, _436_s];
            try {
                for (var conditions_25 = __values(conditions), conditions_25_1 = conditions_25.next(); !conditions_25_1.done; conditions_25_1 = conditions_25.next()) {
                    var condition = conditions_25_1.value;
                    var error = condition(object);
                    if (!error)
                        return null;
                }
            }
            catch (e_41_1) { e_41 = { error: e_41_1 }; }
            finally {
                try {
                    if (conditions_25_1 && !conditions_25_1.done && (_a = conditions_25.return)) _a.call(conditions_25);
                }
                finally { if (e_41) throw e_41.error; }
            }
            return { message: "validation failed at " + path.join(".") + ": there are no valid alternatives", path: path.slice(), reason: { type: "union" } };
        } function _422_s(object) {
            var e_42, _a;
            ;
            if (typeof object !== "object" || object === null || Array.isArray(object))
                return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } };
            {
                if ("file" in object) {
                    path.push("file");
                    var error = su__undefined__string_eu(object["file"]);
                    path.pop();
                    if (error)
                        return error;
                }
            }
            {
                if ("security" in object) {
                    path.push("security");
                    var error = su__undefined_sa_su_si__424__425_ei_si__424__427_ei_si__424__429_ei_si__424__431_ei_eu_ea_433_433_433_eu(object["security"]);
                    path.pop();
                    if (error)
                        return error;
                }
            }
            {
                if ("parameters" in object) {
                    path.push("parameters");
                    var error = su__undefined__436_eu(object["parameters"]);
                    path.pop();
                    if (error)
                        return error;
                }
            }
            try {
                for (var _b = __values(Object.keys(object)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var key = _c.value;
                    if (key !== "file" && key !== "security" && key !== "parameters")
                        return { message: "validation failed at " + path.join(".") + ": " + ("superfluous property '" + key + "' in object"), path: path.slice(), reason: { type: "superfluous-property" } };
                }
            }
            catch (e_42_1) { e_42 = { error: e_42_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_42) throw e_42.error; }
            }
            return null;
        } function su__string__422_eu(object) {
            var e_43, _a;
            var conditions = [_string, _422_s];
            try {
                for (var conditions_26 = __values(conditions), conditions_26_1 = conditions_26.next(); !conditions_26_1.done; conditions_26_1 = conditions_26.next()) {
                    var condition = conditions_26_1.value;
                    var error = condition(object);
                    if (!error)
                        return null;
                }
            }
            catch (e_43_1) { e_43 = { error: e_43_1 }; }
            finally {
                try {
                    if (conditions_26_1 && !conditions_26_1.done && (_a = conditions_26.return)) _a.call(conditions_26);
                }
                finally { if (e_43) throw e_43.error; }
            }
            return { message: "validation failed at " + path.join(".") + ": there are no valid alternatives", path: path.slice(), reason: { type: "union" } };
        } function _378_s(object) {
            var e_44, _a;
            ;
            if (typeof object !== "object" || object === null || Array.isArray(object))
                return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } };
            try {
                for (var _b = __values(Object.keys(object)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var key = _c.value;
                    path.push(key);
                    var error = su__string__422_eu(object[key]);
                    path.pop();
                    if (error)
                        return error;
                }
            }
            catch (e_44_1) { e_44 = { error: e_44_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_44) throw e_44.error; }
            }
            return null;
        } function su__undefined__378_eu(object) {
            var e_45, _a;
            var conditions = [_undefined, _378_s];
            try {
                for (var conditions_27 = __values(conditions), conditions_27_1 = conditions_27.next(); !conditions_27_1.done; conditions_27_1 = conditions_27.next()) {
                    var condition = conditions_27_1.value;
                    var error = condition(object);
                    if (!error)
                        return null;
                }
            }
            catch (e_45_1) { e_45 = { error: e_45_1 }; }
            finally {
                try {
                    if (conditions_27_1 && !conditions_27_1.done && (_a = conditions_27.return)) _a.call(conditions_27);
                }
                finally { if (e_45) throw e_45.error; }
            }
            return { message: "validation failed at " + path.join(".") + ": there are no valid alternatives", path: path.slice(), reason: { type: "union" } };
        } function _375_s(object) {
            var e_46, _a;
            ;
            if (typeof object !== "object" || object === null || Array.isArray(object))
                return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } };
            {
                if ("profiles" in object) {
                    path.push("profiles");
                    var error = su__undefined__376_eu(object["profiles"]);
                    path.pop();
                    if (error)
                        return error;
                }
            }
            {
                if ("providers" in object) {
                    path.push("providers");
                    var error = su__undefined__378_eu(object["providers"]);
                    path.pop();
                    if (error)
                        return error;
                }
            }
            try {
                for (var _b = __values(Object.keys(object)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var key = _c.value;
                    if (key !== "profiles" && key !== "providers")
                        return { message: "validation failed at " + path.join(".") + ": " + ("superfluous property '" + key + "' in object"), path: path.slice(), reason: { type: "superfluous-property" } };
                }
            }
            catch (e_46_1) { e_46 = { error: e_46_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_46) throw e_46.error; }
            }
            return null;
        } var error = _375_s(object); return error; });
    }
    catch (error) {
        if (error instanceof typescript_is_1.TypeGuardError) {
            throw new error_1.AssertionError("Super.json " + error.message, error.path);
        }
        throw error;
    }
    try {
        // NOTE: This is value validation that can't be done by typescript-is now
        //       This is just temporary and will be performed by super.json semantic validator when it's done
        for (var _d = __values(Object.entries(parsedInput.profiles || {})), _e = _d.next(); !_e.done; _e = _d.next()) {
            var _f = __read(_e.value, 2), profileKey = _f[0], profile = _f[1];
            if (typeof profile === 'string') {
                if (!isVersionString(profile) && !isFileURIString(profile)) {
                    throw new error_1.AssertionError('there are no valid alternatives', [
                        '$',
                        'profiles',
                        profileKey,
                    ]);
                }
            }
            else {
                if ('version' in profile && !isVersionString(profile.version)) {
                    throw new error_1.AssertionError('invalid version string format', [
                        '$',
                        'profiles',
                        profileKey,
                        'version',
                    ]);
                }
                try {
                    for (var _g = (e_2 = void 0, __values(Object.entries(profile.providers || {}))), _h = _g.next(); !_h.done; _h = _g.next()) {
                        var _j = __read(_h.value, 2), providerKey = _j[0], provider = _j[1];
                        if (typeof provider === 'string') {
                            if (!isVersionString(provider) && !isFileURIString(provider)) {
                                throw new error_1.AssertionError('there are no valid alternatives', [
                                    '$',
                                    'profiles',
                                    profileKey,
                                    'providers',
                                    providerKey,
                                ]);
                            }
                        }
                    }
                }
                catch (e_2_1) { e_2 = { error: e_2_1 }; }
                finally {
                    try {
                        if (_h && !_h.done && (_b = _g.return)) _b.call(_g);
                    }
                    finally { if (e_2) throw e_2.error; }
                }
            }
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (_e && !_e.done && (_a = _d.return)) _a.call(_d);
        }
        finally { if (e_1) throw e_1.error; }
    }
    try {
        for (var _k = __values(Object.entries(parsedInput.providers || {})), _l = _k.next(); !_l.done; _l = _k.next()) {
            var _m = __read(_l.value, 2), providerKey = _m[0], provider = _m[1];
            if (typeof provider === 'string') {
                if (!isFileURIString(provider)) {
                    throw new error_1.AssertionError('there are no valid alternatives', [
                        '$',
                        'providers',
                        providerKey,
                    ]);
                }
            }
        }
    }
    catch (e_3_1) { e_3 = { error: e_3_1 }; }
    finally {
        try {
            if (_l && !_l.done && (_c = _k.return)) _c.call(_k);
        }
        finally { if (e_3) throw e_3.error; }
    }
    return parsedInput;
};
exports.assertSuperJsonDocument = assertSuperJsonDocument;
exports.isApiKeySecurityValues = typescript_is_1.createIs(function (object) { var path = ["$"]; function _string(object) { ; if (typeof object !== "string")
    return { message: "validation failed at " + path.join(".") + ": expected a string", path: path.slice(), reason: { type: "string" } };
else
    return null; } function _424(object) { ; if (typeof object !== "object" || object === null || Array.isArray(object))
    return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } }; {
    if ("id" in object) {
        path.push("id");
        var error = _string(object["id"]);
        path.pop();
        if (error)
            return error;
    }
    else
        return { message: "validation failed at " + path.join(".") + ": expected 'id' in object", path: path.slice(), reason: { type: "missing-property", property: "id" } };
} return null; } function _425(object) { ; if (typeof object !== "object" || object === null || Array.isArray(object))
    return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } }; {
    if ("apikey" in object) {
        path.push("apikey");
        var error = _string(object["apikey"]);
        path.pop();
        if (error)
            return error;
    }
    else
        return { message: "validation failed at " + path.join(".") + ": expected 'apikey' in object", path: path.slice(), reason: { type: "missing-property", property: "apikey" } };
} return null; } function si__424_s__425_s_ei_s(object) {
    var e_47, _a, e_48, _b;
    var conditions = [_424, _425];
    try {
        for (var conditions_28 = __values(conditions), conditions_28_1 = conditions_28.next(); !conditions_28_1.done; conditions_28_1 = conditions_28.next()) {
            var condition = conditions_28_1.value;
            var error = condition(object);
            if (error)
                return error;
        }
    }
    catch (e_47_1) { e_47 = { error: e_47_1 }; }
    finally {
        try {
            if (conditions_28_1 && !conditions_28_1.done && (_a = conditions_28.return)) _a.call(conditions_28);
        }
        finally { if (e_47) throw e_47.error; }
    }
    try {
        for (var _c = __values(Object.keys(object)), _d = _c.next(); !_d.done; _d = _c.next()) {
            var key = _d.value;
            if (key !== "id" && key !== "apikey")
                return { message: "validation failed at " + path.join(".") + ": " + ("superfluous property '" + key + "' in object"), path: path.slice(), reason: { type: "superfluous-property" } };
        }
    }
    catch (e_48_1) { e_48 = { error: e_48_1 }; }
    finally {
        try {
            if (_d && !_d.done && (_b = _c.return)) _b.call(_c);
        }
        finally { if (e_48) throw e_48.error; }
    }
    return null;
} var error = si__424_s__425_s_ei_s(object); return error; });
exports.isBasicAuthSecurityValues = typescript_is_1.createIs(function (object) { var path = ["$"]; function _string(object) { ; if (typeof object !== "string")
    return { message: "validation failed at " + path.join(".") + ": expected a string", path: path.slice(), reason: { type: "string" } };
else
    return null; } function _424(object) { ; if (typeof object !== "object" || object === null || Array.isArray(object))
    return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } }; {
    if ("id" in object) {
        path.push("id");
        var error = _string(object["id"]);
        path.pop();
        if (error)
            return error;
    }
    else
        return { message: "validation failed at " + path.join(".") + ": expected 'id' in object", path: path.slice(), reason: { type: "missing-property", property: "id" } };
} return null; } function _427(object) { ; if (typeof object !== "object" || object === null || Array.isArray(object))
    return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } }; {
    if ("username" in object) {
        path.push("username");
        var error = _string(object["username"]);
        path.pop();
        if (error)
            return error;
    }
    else
        return { message: "validation failed at " + path.join(".") + ": expected 'username' in object", path: path.slice(), reason: { type: "missing-property", property: "username" } };
} {
    if ("password" in object) {
        path.push("password");
        var error = _string(object["password"]);
        path.pop();
        if (error)
            return error;
    }
    else
        return { message: "validation failed at " + path.join(".") + ": expected 'password' in object", path: path.slice(), reason: { type: "missing-property", property: "password" } };
} return null; } function si__424_s__427_s_ei_s(object) {
    var e_49, _a, e_50, _b;
    var conditions = [_424, _427];
    try {
        for (var conditions_29 = __values(conditions), conditions_29_1 = conditions_29.next(); !conditions_29_1.done; conditions_29_1 = conditions_29.next()) {
            var condition = conditions_29_1.value;
            var error = condition(object);
            if (error)
                return error;
        }
    }
    catch (e_49_1) { e_49 = { error: e_49_1 }; }
    finally {
        try {
            if (conditions_29_1 && !conditions_29_1.done && (_a = conditions_29.return)) _a.call(conditions_29);
        }
        finally { if (e_49) throw e_49.error; }
    }
    try {
        for (var _c = __values(Object.keys(object)), _d = _c.next(); !_d.done; _d = _c.next()) {
            var key = _d.value;
            if (key !== "id" && key !== "username" && key !== "password")
                return { message: "validation failed at " + path.join(".") + ": " + ("superfluous property '" + key + "' in object"), path: path.slice(), reason: { type: "superfluous-property" } };
        }
    }
    catch (e_50_1) { e_50 = { error: e_50_1 }; }
    finally {
        try {
            if (_d && !_d.done && (_b = _c.return)) _b.call(_c);
        }
        finally { if (e_50) throw e_50.error; }
    }
    return null;
} var error = si__424_s__427_s_ei_s(object); return error; });
exports.isBearerTokenSecurityValues = typescript_is_1.createIs(function (object) { var path = ["$"]; function _string(object) { ; if (typeof object !== "string")
    return { message: "validation failed at " + path.join(".") + ": expected a string", path: path.slice(), reason: { type: "string" } };
else
    return null; } function _424(object) { ; if (typeof object !== "object" || object === null || Array.isArray(object))
    return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } }; {
    if ("id" in object) {
        path.push("id");
        var error = _string(object["id"]);
        path.pop();
        if (error)
            return error;
    }
    else
        return { message: "validation failed at " + path.join(".") + ": expected 'id' in object", path: path.slice(), reason: { type: "missing-property", property: "id" } };
} return null; } function _429(object) { ; if (typeof object !== "object" || object === null || Array.isArray(object))
    return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } }; {
    if ("token" in object) {
        path.push("token");
        var error = _string(object["token"]);
        path.pop();
        if (error)
            return error;
    }
    else
        return { message: "validation failed at " + path.join(".") + ": expected 'token' in object", path: path.slice(), reason: { type: "missing-property", property: "token" } };
} return null; } function si__424_s__429_s_ei_s(object) {
    var e_51, _a, e_52, _b;
    var conditions = [_424, _429];
    try {
        for (var conditions_30 = __values(conditions), conditions_30_1 = conditions_30.next(); !conditions_30_1.done; conditions_30_1 = conditions_30.next()) {
            var condition = conditions_30_1.value;
            var error = condition(object);
            if (error)
                return error;
        }
    }
    catch (e_51_1) { e_51 = { error: e_51_1 }; }
    finally {
        try {
            if (conditions_30_1 && !conditions_30_1.done && (_a = conditions_30.return)) _a.call(conditions_30);
        }
        finally { if (e_51) throw e_51.error; }
    }
    try {
        for (var _c = __values(Object.keys(object)), _d = _c.next(); !_d.done; _d = _c.next()) {
            var key = _d.value;
            if (key !== "id" && key !== "token")
                return { message: "validation failed at " + path.join(".") + ": " + ("superfluous property '" + key + "' in object"), path: path.slice(), reason: { type: "superfluous-property" } };
        }
    }
    catch (e_52_1) { e_52 = { error: e_52_1 }; }
    finally {
        try {
            if (_d && !_d.done && (_b = _c.return)) _b.call(_c);
        }
        finally { if (e_52) throw e_52.error; }
    }
    return null;
} var error = si__424_s__429_s_ei_s(object); return error; });
exports.isDigestSecurityValues = typescript_is_1.createIs(function (object) { var path = ["$"]; function _string(object) { ; if (typeof object !== "string")
    return { message: "validation failed at " + path.join(".") + ": expected a string", path: path.slice(), reason: { type: "string" } };
else
    return null; } function _424(object) { ; if (typeof object !== "object" || object === null || Array.isArray(object))
    return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } }; {
    if ("id" in object) {
        path.push("id");
        var error = _string(object["id"]);
        path.pop();
        if (error)
            return error;
    }
    else
        return { message: "validation failed at " + path.join(".") + ": expected 'id' in object", path: path.slice(), reason: { type: "missing-property", property: "id" } };
} return null; } function _431(object) { ; if (typeof object !== "object" || object === null || Array.isArray(object))
    return { message: "validation failed at " + path.join(".") + ": expected an object", path: path.slice(), reason: { type: "object" } }; {
    if ("digest" in object) {
        path.push("digest");
        var error = _string(object["digest"]);
        path.pop();
        if (error)
            return error;
    }
    else
        return { message: "validation failed at " + path.join(".") + ": expected 'digest' in object", path: path.slice(), reason: { type: "missing-property", property: "digest" } };
} return null; } function si__424_s__431_s_ei_s(object) {
    var e_53, _a, e_54, _b;
    var conditions = [_424, _431];
    try {
        for (var conditions_31 = __values(conditions), conditions_31_1 = conditions_31.next(); !conditions_31_1.done; conditions_31_1 = conditions_31.next()) {
            var condition = conditions_31_1.value;
            var error = condition(object);
            if (error)
                return error;
        }
    }
    catch (e_53_1) { e_53 = { error: e_53_1 }; }
    finally {
        try {
            if (conditions_31_1 && !conditions_31_1.done && (_a = conditions_31.return)) _a.call(conditions_31);
        }
        finally { if (e_53) throw e_53.error; }
    }
    try {
        for (var _c = __values(Object.keys(object)), _d = _c.next(); !_d.done; _d = _c.next()) {
            var key = _d.value;
            if (key !== "id" && key !== "digest")
                return { message: "validation failed at " + path.join(".") + ": " + ("superfluous property '" + key + "' in object"), path: path.slice(), reason: { type: "superfluous-property" } };
        }
    }
    catch (e_54_1) { e_54 = { error: e_54_1 }; }
    finally {
        try {
            if (_d && !_d.done && (_b = _c.return)) _b.call(_c);
        }
        finally { if (e_54) throw e_54.error; }
    }
    return null;
} var error = si__424_s__431_s_ei_s(object); return error; });
