// @superfaceai/parser/dist/language/syntax/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseEnvFeatures = exports.isFeature = exports.allFeatures = exports.PARSER_FEATURES = exports.parseRule = exports.parseMap = exports.parseProfile = exports.mapRules = exports.profileRules = exports.rule = exports.parse = void 0;
exports.parse = __importStar(require("./parser"));
exports.rule = __importStar(require("./rule"));
exports.profileRules = __importStar(require("./rules/profile"));
exports.mapRules = __importStar(require("./rules/map"));
var parser_1 = require("./parser");
Object.defineProperty(exports, "parseProfile", { enumerable: true, get: function () { return parser_1.parseProfile; } });
Object.defineProperty(exports, "parseMap", { enumerable: true, get: function () { return parser_1.parseMap; } });
Object.defineProperty(exports, "parseRule", { enumerable: true, get: function () { return parser_1.parseRule; } });
var features_1 = require("./features");
Object.defineProperty(exports, "PARSER_FEATURES", { enumerable: true, get: function () { return features_1.PARSER_FEATURES; } });
Object.defineProperty(exports, "allFeatures", { enumerable: true, get: function () { return features_1.allFeatures; } });
Object.defineProperty(exports, "isFeature", { enumerable: true, get: function () { return features_1.isFeature; } });
Object.defineProperty(exports, "parseEnvFeatures", { enumerable: true, get: function () { return features_1.parseEnvFeatures; } });
